postal.8
.TH "postal" "8" "0.70" "russell@coker.com.au" "Postal"
.SH "NAME"
postal \- program to test SMTP mail server throughput.

.SH "SYNOPSIS"
.B postal
.I [\-m maximum\-message\-size] [\-M minimum\-message\-size] [\-t threads]
.I [\-c messages\-per\-connection] [\-r messages\-per\-minute] [\-s ssl\-percentage]
.I [\-l local\-address] [\-a] [\-b [no]netscape] [\-[z|Z] debug\-file]
.I [\-f sender\-file] smtp\-server user\-list\-filename

.SH "DESCRIPTION"
This manual page documents briefly the
.BR postal ,
program.
.P
It is designed to test the performance of SMTP email servers by sending random
messages to the specified server as fast as possible.
.P
The
.B smtp\-server
parameter specifies the IP address or name of the mail server that the mail
is to be sent to.  Mail sent by
.B Postal
will not use MX records, this is to allow testing outbound relays etc.  If you
want to specify a port other than port 25 then enclose the host address in
square brackets and have the port address immidiately following.  If you want
a DNS lookup for every connection (for testing round\-robin DNS) then
immediately preceed the host address with a '+' character.  To specify multiple
servers for round\-robin use then seperate the addresses with commas.  Note
that
.B localhost
is used for connecting to the same machine.
.P
The
.B user\-list\-filename
is the name of a file which contains a list of user's email addresses.  This
can be just user\-names or fully qualified email addresses.  Whatever you
specify will be sent exactly in the SMTP protocol so make sure you do whatever
is appropriate.  If unsure then use fully qualified addresses (IE
user@example.com).
.P
The
.B sender\-file
contains a list of users that will be in the From: field and envelope sender
of the messages.  If it is not specified then the user\-list\-filename will
be used for the sender list.
.P
The
.B maximum\-message\-size
indicates the size in Kilobytes that will be the maximum size of the
message body.  The size of each message body will be a random number between 0
and the maximum size.  Specify "0" if you want just headers to test the
connection rate.  The default value is 10.
.P
The
.B threads
parameter is the number of threads that should be created to attempt
seperate connections.  A well configured mail server won't accept an unlimited
number of connections so make sure you don't specify a number larger than the
number your mail server is configured to handle.  Also for sensible results
make sure that you don't use enough to make your server thrash as the results
won't be representative of real\-world use.  The default value is 1, this
default is not suitable for real tests, it's just for testing your
configuration.  If you specify multiple server addresses for round\-robin use
then this number of threads will be created per server, IE 4 servers and
.B \-p5
will give 20 threads total.
.P
The
.B messages\-per\-connection
parameter is for sending more than one message per SMTP connection.  The
default value is "1".  A value of \-1 means to send an indefinate number of
messages on one connection (~4 billion).  If a value > 1 is specified then
the number sent on each connection is a random number between 1 and the number
specified.  For simulating a mail server connected directly to the net use a
value of 2 or 3.  For simulating a mail server connected to a front\-end
relay use a large number.  For testing for bugs in your mail server use the
value 0 and leave it running for a week.  ;)  A value of 0 means to disconnect
without sending any messages.  Good for testing a LocalDirector.
.P
The
.B max\-messages\-per\-minute
parameter is for limiting the throughput of the program.  This is designed to
be used when you want to test the performance of other programs when the
system is under load.  The default is effectively 24000 messages per minute.
.P
The
.B local\-address
parameter specifies which local IP address(es) are used to make the outbound
connections.  Specified in the same way as the remote address.  This is good
for testing LocalDirectors or other devices that perform differently depending
on which source IP address was used.
.P
The
.B \-a
command turns on all logging.  All message data received will be logged.  This
will make it slow and it may not be able to saturate a fast Ethernet link...
.P
The
.B \-b
switch allows you to specify breakage strings.  Currently the only option is
for Netscape mail server which strips spaces from the start of subject lines.
.B \-b netscape
means to avoid leading spaces on subject fields to not break Netscape.
.B \-b nonetscape
means to always put extra space to test for the bug in Netscape and similar
products.  Some people say that the RFCs are open to interpretation on this
issue, I am interested to see whether anyone else interprets it the way that
Netscape does.
.P
The
.B \-s
switch specifies the percentage of connections which are to use
.B TLS
AKA
.B SSL.
Use 0 for no SSL, or 100 for always SSL, or any number in between.  Default is
0.
.P
The
.B \-z
switch allows you to specify a debugging file base.  From this base one file
is created for each thread (with a ':' and the thread number appended), each
file is used to log all IO performed by that thread for debugging purposes.
.P
The
.B \-Z
switch is the same but creates a separate file for each connection as well
with an attitional ':' appended followed by the connection number.

.SH "NOTES"

When testing mail servers please use domains defined in http://www.rfc-editor.org/rfc/rfc2606.txt - example.com, example.net, and example.org are all good options.  Please don't use anything related to a valid name, that will cause pain for you and others on the net.

.SH "RETURN CODES"
.TP
.B 0
No Error
.TP
.B 1
Bad Parameters
.TP
.B 2
System Error, lack of memory or some other resource

.SH "AUTHOR"
This program, it's manual page, and the Debian package were written by
Russell Coker <russell@coker.com.au>.

.SH "AVAILABILITY"
The source is available from http://doc.coker.com.au/projects/postal/ .
.P
See http://etbe.coker.com.au/category/benchmark for further information.

.SH "SEE ALSO"
.BR bhm (8), postal\-list (8), rabid (8)


debian/changelog
postal (0.72) unstable; urgency=low

  * Make LMTP work.
  * Accept a tab as a field delimiter for the user-list file.

 -- Russell Coker <russell@coker.com.au>  Sat, 14 Jan 2012 19:27:21 +1100

postal (0.71) unstable; urgency=low

  * Made rabid get the password right when -u is used.  Thanks to Lars Nordin
    for the patch.
  * Made rabid correctly check for the '.' at the end of the results of the
    list command.
  * Made all the programs correctly handle a userlist that has empty lines.
  * Use the -lgcrypt option when using gcrypt.
  * Only call gcry_control() if GNUTLS_VERSION_NUMBER <= 0x020b00
  * Ignore blank lines in user list.
  * Add subject, date, and msgid to the md5sum in the client.
    Closes: #611907
  * Stop using gnutls_kx_set_priority()
    Closes: #624038
  * Added a patch from Michael Brown <michael@netdirect.ca> to make an RPM build
    on RHEL5 (and probably other RPM systems).
  * Added a patch from Brad Hards <bradh@frogmouth.net> to fix typos in the
    man page.
  * Rewrote a lot of the md5 checking code, now the md5 hash is stored in a
    header.  It should work now.

 -- Russell Coker <russell@coker.com.au>  Mon, 02 Jan 2012 17:08:50 +1100

postal (0.70) unstable; urgency=low

  * Made it build correctly when no SSL support is configured.  Also tidy
    up the documentation.

 -- Russell Coker <russell@coker.com.au>  Mon, 26 May 2008 19:33:48 +1000

postal (0.69) unstable; urgency=low

  * Really fix the GCC 4.3 issues.  The main change was to remove the
    OS/2 and Win32 support (which probably didn't work anyway).  The previous
    version actually fixed GCC 4.2 issues.  I don't plan to fix the
    ext/hash_map issue yet, it doesn't stop anything from working and I'm
    afraid of breaking things for the majority of systems which use older
    headers.  Thanks to Bradley Smith and Chris Lamb for sending a patch.
    Closes: #474831
    Thanks to Martin Michlmayr for sending a patch.
    Closes: #417480

 -- Russell Coker <russell@coker.com.au>  Sat, 19 Apr 2008 08:07:00 +1000

postal (0.68) unstable; urgency=low

  * Added $(DESTDIR) support to the Makefile and -lrt for nanosleep() on
    OpenSolaris (the OpenSolaris code is not tested).
  * Fixed a minor compilation bug when not using GNUTLS.
    Thanks Frank Fegert <frank.fegert@web.de>.
  * Made it not use C++ warning flags on C compilation.
  * Updated it to the latest Debian standards for Lenny and compile with
    GCC 4.3. Closes: #474831

 -- Russell Coker <russell@coker.com.au>  Wed,  9 Apr 2008 11:16:00 +1000

postal (0.67-0.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Fix FTBFS with gcc-4.3. Thanks to MArtin Michlmayr. Closes: #417480

 -- Andreas Barth <aba@not.so.argh.org>  Sun, 16 Mar 2008 23:48:44 +0000

postal (0.67) unstable; urgency=low

  * Changed license to GPLv3.

 -- Russell Coker <russell@coker.com.au>  Mon,  2 Jul 2007 20:12:00 +1000

postal (0.66) unstable; urgency=low

  * Added MessageId to Postal.
  * Added "make dep".
  * Made GNUTLS work in bhm.

 -- Russell Coker <russell@coker.com.au>  Fri,  2 Mar 2007 18:44:00 +1100

postal (0.65) unstable; urgency=low

  * Added new program bhm to listen on port 25 and send mail to /dev/null.
    This allows testing mail relay systems.
  * Fixed a minor bug in reporting when compiled without SSL.
  * Made postal write the date header in correct RFC2822 format.
  * Removed the name-expansion feature, it confused many people and is not
    needed now that desktop machines typically have 1G of RAM.
    Now postal and rabid can have the same user-list file.
  * Moved postal-list into the bin directory.
  * Changed the thread stack size to 32K (used to be the default of 10M) to
    save virtual memory size (not that this makes much difference to
    anything other than the maximum number of threads on i386).
  * Added a minimum message size option to Postal.
  * Added a Postal option to specify a list of sender addresses separately to
    the list of recipient addresses.
  * Removed some unnecessary error messages.
  * Handle EINTR to allow ^Z and "bg" from the command line.  I probably
    don't handle all cases, but now that I agree that failure to handle ^Z is
    an error I expect bug reports.
  * Made the test programs display output on the minute, previously they
    displayed once per minute (EG 11:10:35) while now it will be 11:10:00.
    This also means that the first minute reported will have something less
    than 60 seconds of data - this does not matter as a mail server takes
    longer than that to get up to speed.
  * Added support for GNUTLS and made the Debian package build with it.
    Note that BHM doesn't yet work correctly with TLS.
  * Made the programs exit cleanly.

 -- Russell Coker <russell@coker.com.au>  Thu, 18 Jan 2007 08:14:00 +1100

postal (0.62.1) unstable; urgency=low

  * NMU
  * Build-Depend: libssl-dev, not libssl096-dev.  Closes: #176956

 -- LaMont Jones <lamont@debian.org>  Sun, 25 May 2003 21:33:07 -0600

postal (0.62) unstable; urgency=high

  * Added an INSTALL document.

  * Made it compile with GCC 3.2.

 -- Russell Coker <russell@coker.com.au>  Tue, 14 Jan 2003 13:28:00 +0100

postal (0.61-1) unstable; urgency=high

  * Changed the section to main.

 -- Russell Coker <russell@coker.com.au>  Sun, 16 Jun 2002 23:17:00 +0200

postal (0.61) unstable; urgency=high

  * Fixed the makefile to not use "-d" or "-p" options to $(INSTALL).

  * Changed connect() functions to Connect() because of some problems with
    Solaris.

  * Removed the optind=0 which caused problems on BSD.

 -- Russell Coker <russell@coker.com.au>  Sun, 16 Jun 2002 21:16:00 +0200

postal (0.60) unstable; urgency=high

  * Added #include <stdlib.h> to mutex.cpp to make exit() be resolved correctly.

  * Fixed the delete percentage code for rabid.  The option "-d100:0" will mean
    that no messages get deleted.

  * Added a new version of the thread code from Bonnie++, fixes a few bugs.


postal.cpp
#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif

#include "userlist.h"
#include "smtp.h"
#include <unistd.h>
#include <sys/wait.h>
#include <signal.h>
#include <stdio.h>
#include "postal.h"
#include "logit.h"
#ifdef USE_GNUTLS
#include <errno.h>
#include <gcrypt.h>
GCRY_THREAD_OPTION_PTHREAD_IMPL;
#endif

void usage()
{
  printf("Usage: postal [-m maximum-message-size] [-M minimum-message-size] [-t threads]\n"
         "              [-c messages-per-connection] [-r messages-per-minute] [-a]\n"
         "              [-b [no]netscape] [-p port] [-[z|Z] debug-file]\n"
#ifdef USE_SSL
         "              [-s ssl-percentage]\n"
#endif
         "              [-L] [-l local-address] [-f sender-file]\n"
         "              smtp-server user-list-filename\n"
         "\n"
         "Postal Version: " VER_STR "\n");
  exit(eParam);
}

int exitCount = 0;

void endit(int)
{
  exitCount++;
  if(exitCount > 2)
    exit(1);
}

int main(int argc, char **argv)
{
  int maxMsgSize = 10;
  int minMsgSize = 0;
  int processes = 1;
  int msgsPerMinute = 0; // 0 for unlimited
  int msgsPerConnection = 1;
  const char *ourAddr = NULL;
  bool allLog = false;
  TRISTATE netscape = eNONE;
  PCCHAR debugName = NULL;
  PCCHAR senderFile = NULL;
  bool debugMultipleFiles = false;
  bool useLMTP = false;
#ifdef USE_SSL
  int ssl = 0;
#endif
#ifdef USE_GNUTLS
#if GNUTLS_VERSION_NUMBER <= 0x020b00
  gcry_control(GCRYCTL_SET_THREAD_CBS, &gcry_threads_pthread);
#endif
  gnutls_global_init();
  if(!gnutls_check_version(GNUTLS_VER))
  {
    fprintf(stderr, "Needs version " GNUTLS_VER " of GNUTLS\n");
    exit(1);
  }
#endif
  unsigned short port = 25;

  int c;
  while(-1 != (c = getopt(argc, argv, "ab:f:m:M:p:s:t:c:r:Ll:z:Z:")) )
  {
    switch(char(c))
    {
      case '?':
      case ':':
        usage();
      break;
      case 'a':
        allLog = true;
      break;
      case 'b':
        if(!strcasecmp(optarg, "netscape"))
          netscape = eWONT;
        else if(!strcasecmp(optarg, "nonetscape"))
          netscape = eMUST;
      break;
      case 'f':
          senderFile = optarg;
      break;
      case 'L':
        useLMTP = true;
      break;
      case 'l':
        ourAddr = optarg;
      break;
      case 'm':
        maxMsgSize = atoi(optarg);
      break;
      case 'M':
        minMsgSize = atoi(optarg);
      break;
      case 'p':
        port = atoi(optarg);
      break;
      case 's':
#ifdef USE_SSL
        ssl = atoi(optarg);
#else
        usage();
#endif
      break;
      case 't':
        processes = atoi(optarg);
      break;
      case 'c':
        msgsPerConnection = atoi(optarg);
      break;
      case 'r':
        msgsPerMinute = atoi(optarg);
      break;
      case 'Z':
        debugMultipleFiles = true;
      case 'z':
        debugName = optarg;
      break;
    }
  }
  if(minMsgSize < 0 || maxMsgSize > MAX_MSG_SIZE || maxMsgSize < minMsgSize)
    usage();

  if(processes < 1 || processes > MAX_PROCESSES)
    usage();
  if(msgsPerMinute < 0 || msgsPerConnection < -1 )
    usage();
#ifdef USE_SSL
  if(ssl < 0 || ssl > 100)
    usage();
#endif

  if(optind + 2 != argc)
    usage();

  UserList ul(argv[optind + 1], false);
  UserList *senderList = NULL;
  if(senderFile)
    senderList = new UserList(senderFile, false);

  int fd[2];
  if(pipe(fd))
  {
    printf("Can't create pipe.\n");
    return eSystem;
  }
  struct sigaction sa;
  sa.sa_sigaction = NULL;
  sa.sa_handler = SIG_IGN;
  sa.sa_flags = 0;
  if(sigaction(SIGPIPE, &sa, NULL))
  {
    printf("Can't block SIGPIPE.\n");
    return eSystem;
  }

  sa.sa_flags = SA_SIGINFO;
  sa.sa_handler = &endit;
  if(sigaction(SIGINT, &sa, NULL))
  {
    printf("Can't handle SIGINT.\n");
    return eSystem;
  }
  
  printf("time,messages,data(K),errors,connections"
#ifdef USE_SSL
    ",SSL connections"
#endif
    "\n");

  Logit log("postal.log", allLog, false, 0);
  Logit *debug = NULL;

  if(debugName)
    debug = new Logit(debugName, false, debugMultipleFiles, 0);

  smtp mailer(&exitCount, argv[optind], ourAddr, ul, senderList, minMsgSize
            , maxMsgSize, msgsPerConnection, processes, &log, netscape, useLMTP
#ifdef USE_SSL
            , ssl
#endif
            , port, debug);

  return mailer.doAllWork(msgsPerMinute);
}

